File: apps/parser/src/integrations/vk.ts
import { URLSearchParams } from 'url';

import { Injectable, OnApplicationBootstrap, OnModuleInit } from '@nestjs/common';
import { PrismaService } from '@tsuwari/prisma';


export interface VKDBData {
  userId: string;
}

interface IVkResponse {
  error?: unknown;
  response?: {
    text?: string;
    audio?: {
      artist: string;
      title: string;
    };
  };
}

@Injectable()
export class VkIntegration implements OnModuleInit {
  #accessToken: string;

  constructor(private readonly prisma: PrismaService) { }

  async onModuleInit() {
    const service = await this.prisma.integration.findFirst({
      where: {
        service: 'VK',
      },
    });

    if (service && service.accessToken) {
      this.#accessToken = service.accessToken;
    }

    return this;
  }

  async fetchSong(vkUserId: string) {
    const params = {
      access_token: this.#accessToken,
      uid: vkUserId,
      v: '5.131',
    };

    const request = await fetch(`https://api.vk.com/method/status.get?${new URLSearchParams(params)}`, {
      method: 'GET',
    });

    const response = await request.json() as IVkResponse;

    if (response.error) return null;
    if (!response.response || !response.response.audio) return null;

    return `${response.response.audio?.artist} — ${response.response.audio?.title}`;
  }
}

File: apps/parserv2/internal/integrations/vk/vk.go
package vk

import (
	"encoding/json"
	"fmt"
	model "tsuwari/parser/internal/models"

	req "github.com/imroc/req/v3"
)

type DbData struct {
	UserId *string `json:"userId"`
}

type Vk struct {
	DbData
	integration *model.ChannelInegrationWithRelation
}

func New(integration *model.ChannelInegrationWithRelation) *Vk {
	if integration == nil || !integration.Data.Valid || !integration.Integration.APIKey.Valid {
		return nil
	}

	dbData := DbData{}
	err := json.Unmarshal([]byte(integration.Data.String), &dbData)
	if err != nil || dbData.UserId == nil {
		return nil
	}

	service := Vk{
		DbData:      dbData,
		integration: integration,
	}

	return &service
}

type VkError struct {
	Code int    `json:"error_code,omitempty"`
	Msg  string `error_msg:"msg,omitempty"`
}

type VkAudio struct {
	Artist string `json:"artist,omitempty"`
	Title  string `json:"title,omitempty"`
}

type VkStatus struct {
	Text  *string  `json:"text,omitempty"`
	Audio *VkAudio `json:"audio,omitempty"`
}

type VkResponse struct {
	Error  *VkError  `json:"error,omitempty"`
	Status *VkStatus `json:"response"`
}

func (c *Vk) GetTrack() *string {
	data := VkResponse{}
	var response string

	resp, err := req.R().
		SetQueryParam("access_token", c.integration.Integration.APIKey.String).
		SetQueryParam("uid", *c.UserId).
		SetQueryParam("v", "5.131").
		SetResult(&data).
		SetContentType("application/json").
		Get("https://api.vk.com/method/status.get")

	if err != nil || !resp.IsSuccess() {
		return nil
	}

	if data.Error != nil || data.Status == nil || data.Status.Audio == nil {
		return nil
	}

	status := *data.Status.Audio
	response = fmt.Sprintf(
		"%s — %s",
		status.Artist,
		status.Title,
	)

	return &response
}
